mdsshr/MdsEvents.c

```diff
@@ -366,20 +383,63 @@ static void KillHandler()
   external_thread_created = 0;
 }
 
-static void *handleRemoteAst(void *arg __attribute__((unused)))
+static void *handleRemoteAst(void *arg )
 {
   INIT_STATUS;
   char buf[16];
-  int i;
+  int idx, i;
   Message *m;
   int selectstat;
   fd_set readfds;
+  int receive_thread_ids[256];
+  int receive_thread_sockets[256];
+  struct MdsipEventInfo *eventInfo = arg;
+  int curr_eventid;
+
+  getServerDefinition("mds_event_server", receive_servers, &num_receive_servers);
+  newRemoteId(eventInfo->eventid);
+  for (idx = 0; idx < num_receive_servers; idx++)
+  {  
+    receive_thread_ids[idx] = searchOpenServer(receive_servers[idx]);
+    if(receive_thread_ids[idx] < 0)
+    {
+      receive_thread_ids[idx] = ConnectToMds_(receive_servers[idx]);
+    }
+    if ( receive_thread_ids[idx] < 0)
+    {
+      printf("\nError connecting to %s", receive_servers[idx]);
+      perror("ConnectToMds_");
+      receive_thread_sockets[idx] = 0;
+    }
+    else
+    {
+      receive_thread_sockets[idx] = 0;
+      GetConnectionInfo_(receive_thread_ids[idx], 0, &receive_thread_sockets[idx], 0);
+    }
+    if (receive_thread_ids[idx] >= 0)
+    {
+      status = MdsEventAst_(receive_thread_ids[idx], eventInfo->eventnam, eventInfo->astadr, eventInfo->astprm,
+                              &curr_eventid);
+      if (STATUS_OK)
+        setRemoteId(*eventInfo->eventid, idx, curr_eventid);
+      else
+      {
+        printf("\nError issuing Event Ast remote %s", receive_servers[idx]);
+        return NULL;
+      }
+    }
+  }
   while (1)
   {
     FD_ZERO(&readfds);
+
     for (i = 0; i < num_receive_servers; i++)
-      if (receive_sockets[i])
-        FD_SET(receive_sockets[i], &readfds);
+    {
```

Comment:
waht was the motivaton to switch from receive_sockets to receive_thread_sockets .. copy and paste or was this actually a bug

Reply:
Now the thread must open the connection itself, this is the reason of the receive_thread_ids and receive_thread_sockets that are private to the thread. Formerly the connection was opened by the main thread but connection structures were not visible by the created thread.

Reply:
I see. hmm i am not sure if our test cover it but i remember for backward compatibility it was important that the main thread holds the threads.. i think it was because jobs may run in a thread and dynamically open a connection.. which would then be closed once the job is done. resulting in very slow network performance.. this may be addressed elsewhere no (treeshr was altered in the way it connects to remote trees) but it should be verified. its quite a while ago.. but i f remember correctly some older mdsip serverse.g. do not multiple connections and hence it is required to used a shared one.

Reply:
I see. hmm i am not sure if our test cover it but i remember for backward compatibility it was important that the main thread holds the threads.. i think it was because jobs may run in a thread and dynamically open a connection.. which would then be closed once the job is done. resulting in very slow network performance.. this may be addressed elsewhere no (treeshr was altered in the way it connects to remote trees) but it should be verified. its quite a while ago.. but if i remember correctly some older mdsip serverse.g. do not multiple connections and hence it is required to used a shared one.

Reply:
that may be but mdsip is not just used for event.. you would break access to old server hosting mdsplus trees

Reply:
I don't know if such mdsip servers still exist, i.e. mdsip servers that do not accept more than one connection, I have no memory of such a limitation.

Reply:
And in any case a single connection is established in MdsEventAst